linera-chain/src/data_types.rs

```diff
@@ -587,19 +587,71 @@ impl<'a> SignatureAggregator<'a> {
     }
 }
 
+fn is_ordered(values: &[(ValidatorName, Signature)]) -> bool {
+    values.windows(2).all(|pair| pair[0].0 <= pair[1].0)
+}
+
+fn has_duplicates(values: &[(ValidatorName, Signature)]) -> bool {
+    let mut encountered = HashSet::new();
+
+    for &(validator_name, _) in values {
+        if !encountered.insert(validator_name) {
+            return true; // Found a duplicate
+        }
+    }
+
+    false // No duplicates found
+}
+
+impl<'de> Deserialize<'de> for Certificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Debug, Deserialize)]
+        #[serde(rename = "Certificate")]
+        struct CertificateHelper {
+            value: HashedValue,
+            round: RoundNumber,
+            signatures: Vec<(ValidatorName, Signature)>,
+        }
+
+        let helper: CertificateHelper = Deserialize::deserialize(deserializer)?;
+        if !is_ordered(&helper.signatures) {
+            Err(serde::de::Error::custom("Vector is not sorted"))
+        } else if has_duplicates(&helper.signatures) {
+            Err(serde::de::Error::custom("Vector has duplicates"))
+        } else {
+            Ok(Self {
+                value: helper.value,
+                round: helper.round,
+                signatures: helper.signatures,
+            })
+        }
+    }
+}
+
 impl Certificate {
     pub fn new(
         value: HashedValue,
         round: RoundNumber,
-        signatures: Vec<(ValidatorName, Signature)>,
+        mut signatures: Vec<(ValidatorName, Signature)>,
     ) -> Self {
+        if !is_ordered(&signatures) {
+            signatures.sort_by_key(|&(validator_name, _)| validator_name)
+        }
+
         Self {
             value,
             round,
             signatures,
         }
     }
 
+    pub fn get_signatures(&self) -> &Vec<(ValidatorName, Signature)> {
```

Comment:
It's idiomatic to omit the `get` in getters:
```suggestion
    pub fn signatures(&self) -> &Vec<(ValidatorName, Signature)> {
```